File: C++/93. Restore IP Addresses.cpp
/*
回溯：横向遍历end，纵向递归start
*/
class Solution {
public:
    int numPoint = 0;
    vector<string> res;
    bool isValid(const string& s, int start, int end) {
        if (start > end) return false;
        if (s[start] == '0' && end > start) return false;
        string temp = s.substr(start, end - start + 1);
        if (temp.size() > 3) return false;
        if (stoi(temp) >= 0 && stoi(temp) <= 255) return true;
        return false;
    }
    void backtrack(string& s, int start) {
        if (numPoint == 3) {
            if (isValid(s, start, s.size() - 1)) {
                res.push_back(s);
            }
            return;
        }
        for (int end = start; end < start + 3 && end < s.size(); end++) {
            if (isValid(s, start, end)) {
                s.insert(s.begin() + end + 1, '.');
                numPoint += 1;
                backtrack(s, end + 2);
                numPoint -= 1;
                s.erase(s.begin() + end + 1);
            }
            else {
                return;
            }
        }
    }
    vector<string> restoreIpAddresses(string s) {
        numPoint = 0;
        res.clear();
        backtrack(s, 0);
        return res;
    }
};

File: Python/93. Restore IP Addresses.py
'''
回溯：横向遍历end，纵向递归start
'''


class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        res = []
        self.numPoint = 0

        def isValid(s, start, end):
            if start > end: return False
            if s[start] == '0' and start < end: return False
            temp = s[start:end + 1]
            if len(temp) > 3: return False
            if int(temp) >= 0 and int(temp) <= 255: return True
            return False

        def backtrack(s, start):
            if self.numPoint == 3 and isValid(s, start, len(s) - 1):
                res.append(s[:])
                return
            if self.numPoint > 3:
                return
            for end in range(start, len(s)):
                if isValid(s, start, end):
                    s = s[:end + 1] + '.' + s[end + 1:]
                    self.numPoint += 1
                    backtrack(s, end + 2)
                    self.numPoint -= 1
                    s = s[:end + 1] + s[end + 2:]
                else:
                    return

        backtrack(s, 0)
        return res
